src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,962 @@
+                                   uint8_t,
+                                   uint16_t,
+                                   char,
+                                   char>;
+  constexpr static layout_type cell_layout(uint8_t unshared_len,
+                                           uint16_t onode_len) {
+    return layout_type{1, 1, 1, unshared_len, onode_len};
+  }
+};
+
+struct tag_t {
+  template<int N, ntype_t node_type>
+  static constexpr tag_t create() {
+    static_assert(std::clamp(N, 0, 3) == N);
+    return tag_t{N, static_cast<uint8_t>(node_type)};
+  }
+  bool is_leaf() const {
+    return ntype_t{node_type} == ntype_t::leaf;
+  }
+  int layout() const {
+    return layout_type;
+  }
+  int layout_type : 4;
+  uint8_t node_type : 4;
+};
+
+// for calculating size of variable-sized item/key
+template<class T>
+size_t size_of(const T& t) {
+  using decayed_t = std::decay_t<T>;
+  if constexpr (std::is_scalar_v<decayed_t>) {
+    return sizeof(decayed_t);
+  } else {
+    return t.size();
+  }
+}
+
+enum class size_state_t {
+  okay,
+  underflow,
+  overflow,
+};
+
+// layout of a node
+template<size_t BlockSize,
+         int N,
+         ntype_t NodeType>
+struct node_t {
+  static_assert(std::clamp(N, 0, 3) == N);
+  constexpr static ntype_t node_type = NodeType;
+  constexpr static int node_n = N;
+
+  using key_t = FixedKey<N, NodeType>;
+  using item_t = std::conditional_t<NodeType == ntype_t::leaf,
+                                    onode_t,
+                                    child_addr_t>;
+  using const_item_t = std::conditional_t<NodeType == ntype_t::leaf,
+                                          const onode_t&,
+                                          child_addr_t>;
+  // XXX: magic numbers here
+  static constexpr bool item_in_key = (N == 3 &&
+                                       NodeType == ntype_t::inner);
+  using partial_key_t = std::conditional_t<N < 3,
+                                           variable_sized_key_t,
+                                           empty_key_t>;
+
+  std::pair<const key_t&, const partial_key_t&>
+  key_at(unsigned slot) const;
+
+  // update an existing oid with the specified item
+  ghobject_t update_oid_with_slot(unsigned slot, const ghobject_t& oid) const;
+  const_item_t item_at(const key_t& key) const;
+  void dump(std::ostream& os) const;
+
+  // for debugging only.
+  static constexpr bool is_leaf() {
+    return node_type == ntype_t::leaf;
+  }
+
+  bool _is_leaf() const {
+    return tag.is_leaf();
+  }
+
+  char* from_end(uint16_t offset);
+  const char* from_end(uint16_t offset) const;
+  uint16_t used_space() const;
+  uint16_t free_space() const {
+    return capacity() - used_space();
+  }
+  static uint16_t capacity();
+  // TODO: if it's allowed to update 2 siblings at the same time, we can have
+  //       B* tree
+  static constexpr uint16_t min_size();
+
+  // @return <minimum bytes to grab, maximum bytes to grab>
```

Comment:
From the implementations, I infer that bytes_to_grab are the allowable bounds on bytes to add to an underfull node and that bytes_to_push returns the allowable bounds on bytes to remove from an overfull node.  I don't think "grab" and "push" really convey that distinction, however.  Other names might help, or you could expand the comments here.

Reply:
i renamed them to `bytes_to_remove` and `bytes_to_add` respectively. and expanded the comments.